source/auth/Credentials.cpp

```diff
@@ -267,6 +269,45 @@ namespace Aws
                 return s_CreateWrappedProvider(
                     aws_credentials_provider_new_chain_default(allocator, &raw_config), allocator);
             }
+
+            std::shared_ptr<ICredentialsProvider> CredentialsProvider::CreateCredentialsProviderX509(
+                const CredentialsProviderX509Config &config,
+                Allocator *allocator)
+            {
+                struct aws_credentials_provider_x509_options raw_config;
+                AWS_ZERO_STRUCT(raw_config);
+
+                raw_config.bootstrap = config.Bootstrap->GetUnderlyingHandle();
+                raw_config.tls_connection_options = config.TlsOptions.GetUnderlyingHandle();
+                raw_config.thing_name = aws_byte_cursor_from_c_str(config.ThingName.c_str());
+                raw_config.role_alias = aws_byte_cursor_from_c_str(config.RoleAlias.c_str());
+                raw_config.endpoint = aws_byte_cursor_from_c_str(config.Endpoint.c_str());
+
+                struct aws_http_proxy_options proxy_options;
+                AWS_ZERO_STRUCT(proxy_options);
+                if (config.ProxyOptions.has_value())
+                {
+                    const Http::HttpClientConnectionProxyOptions &proxy_config = config.ProxyOptions.value();
+
+                    proxy_options.host = aws_byte_cursor_from_c_str(proxy_config.HostName.c_str());
+                    proxy_options.port = proxy_config.Port;
+                    proxy_options.tls_options = proxy_config.TlsOptions->GetUnderlyingHandle();
+                    proxy_options.auth_type = (enum aws_http_proxy_authentication_type)proxy_config.AuthType;
+                    proxy_options.auth_username = aws_byte_cursor_from_c_str(proxy_config.BasicAuthUsername.c_str());
```

Comment:
Some of these are optional right? Should we do a check that they’ve been set?

Reply:
username and password being null will just result in 0-length cursors.  Everything else is required.